valhalla/common/telescope_states.py

```diff
@@ -103,108 +119,83 @@ def _get_es_data(self, sites, telescopes):
+                        current_lump = self._create_lump(telcode, event_type, event_reason, event_start)
+                else:
+                    current_lump = self._create_lump(telcode, event_type, event_reason, event_start)
 
-        # If the event is close enough to the latest timestamp in the lump, it belongs in that lump.
-        if time_diff < dt:
-            return True
+        if current_lump['telescope']:
+            # We have a final current lump we were in, so save it
+            self._save_lump(telescope_states, current_lump, self.end)
 
-        # If the event is far in time from the lump, but has the same reason, that means it came from the same
-        # scheduling run as that timestamp and belongs in the lump.
-        if time_diff >= dt and event_source['reason'] in lump_data['reasons']:
-            return True
+        return telescope_states
 
-        # If the event is an ENCLOSURE_INTERLOCK, and the lump includes a SEQUENCER_UNAVAILABLE, it is part of the
-        # lump. This is because the two states are redundant.
-        if event_source['type'] == 'ENCLOSURE_INTERLOCK' and 'SEQUENCER_UNAVAILABLE' in lump_data['types']:
-            return True
+    def _save_lump(self, telescope_states, lump, end):
+        lump['end'] = min(self.end, end)
+        lump['start'] = max(self.start, lump['start'])
+        telkey = lump['telescope']
+        lump['telescope'] = str(lump['telescope'])
+        if telkey not in telescope_states:
+            telescope_states[telkey] = []
+        telescope_states[telkey].append(lump)
 
-        return False
+        return telescope_states
 
     @staticmethod
-    def _categorize(lump):
-        # TODO: Categorize each lump in a useful way for network users.
-        event_type, event_reason = lump['types'][0], lump['reasons'][0]
-        for this_type, this_reason in zip(lump['types'], lump['reasons']):
-            # If the state in ENCLOSURE INTERLOCK, wait to instead categorize as SEQUENCER_UNAVAILABLE.
-            if this_type.upper() != 'ENCLOSURE_INTERLOCK':
-                event_type, event_reason = this_type, this_reason
-                break
-        return event_type, event_reason
-
-    def _lump_end(self, lump, next_event=None):
-        if not next_event or str(lump['telescope']) != str(self._telescope(next_event)):
-            return self.end
-        return string_to_datetime(next_event['timestamp'])
-
-    def _set_lump(self, event):
+    def _create_lump(telcode, type, reason, start):
         return {
-            'reasons': [event['_source']['reason']],
-            'types': [event['_source']['type']],
-            'start': string_to_datetime(event['_source']['timestamp']),
-            'telescope': self._telescope(event['_source']),
-            'latest_timestamp': string_to_datetime(event['_source']['timestamp'])
+            'telescope': telcode,
+            'event_type': type,
```

Comment:
Try to avoid using builtins whenever possible.

Reply:
good idea